Dockerfile

```diff
@@ -53,6 +61,8 @@ RUN apt-get update \
 	&& apt-get -y install --no-install-recommends gcc wget default-jre-headless \
 	&& cd /build \
 	&& make install \
-	&& make test \
+	# test always fail, resources not available for download. Resources should be made available
+	# somewhere else, e.g. github.com/OCR-D/assets
+	# && make test \
```

Comment:
Please revert this part – I'll try to get the tests working again. 

(The problem is merely that OCR-D GT has moved, and the Ocropy [models](http://cis.lmu.de/~finkf/fraktur1-00085000.pyrnn.gz) trained by @finkf are not downloadable anymore – but for that test, any pair of models would be sufficient AFAICS.)

@tboenig, would you happen to know which GT bag now belongs to `https://ocr-d-repo.scc.kit.edu/api/v1/dataresources/75ad9f94-dbaa-43e0-ab06-2ce24c497c61/data`?

Reply:
done. Don't know if it actually works – TravisCI is not available anymore. I created a CircleCI config, but would need permissions to set it up on your fork. @joschrew can you please [set up a CircleCI project for this](https://circleci.com/docs/getting-started/)?

Reply:
Thanks!

Unfortunately, I have no permissions to look into the [new CI status](https://app.circleci.com/pipelines/circleci/3UxtJNDEXR2FBFZHw8XvMG/UW3jxqgQiV4Nnz1CGo8HyJ/3/workflows/2c7976af-7e08-44e0-97cc-42462caa3c5e) – no idea what went wrong.

Since it is still failing: could you please post the relevant error messages? (I suspect my updated GT URLs are the culprit.)

Reply:
I am working on it. The test still tries to download from non existing kit. I think I found out which resources where used and I am now trying to locate them and change the test.

Reply:
I have now changed the `make test` output to print everything. This is what I get. To me it seems the downloading works because the first test passes (IIUC). I don't have an idea what this error is about.

<details>
<summary> Error msg </summary>
</details>

```
bash tests/run_add_zip_test.bash 
--2024-02-13 14:12:37--  http://hdl.handle.net/21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780
Resolving hdl.handle.net (hdl.handle.net)... 52.32.51.238, 52.19.202.222, 54.163.52.124, ...
Connecting to hdl.handle.net (hdl.handle.net)|52.32.51.238|:80... connected.
HTTP request sent, awaiting response... 302 
Location: http://ola-hd.ocr-d.de/export?id=21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780&internalId=false [following]
--2024-02-13 14:12:38--  http://ola-hd.ocr-d.de/export?id=21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780&internalId=false
Resolving ola-hd.ocr-d.de (ola-hd.ocr-d.de)... 141.5.99.45
Connecting to ola-hd.ocr-d.de (ola-hd.ocr-d.de)|141.5.99.45|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ola-hd.ocr-d.de/export?id=21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780&internalId=false [following]
--2024-02-13 14:12:38--  https://ola-hd.ocr-d.de/export?id=21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780&internalId=false
Connecting to ola-hd.ocr-d.de (ola-hd.ocr-d.de)|141.5.99.45|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘/tmp/tmp.oJvWETFBhO/download/blumenbach_anatomie_1805.ocrd.zip’

/tmp/tmp.oJvWETFBhO     [                <=> ]  41.18M  12.4MB/s    in 3.3s    

2024-02-13 14:12:42 (12.4 MB/s) - ‘/tmp/tmp.oJvWETFBhO/download/blumenbach_anatomie_1805.ocrd.zip’ saved [43176289]

14:12:43.920 INFO ocrd.workspace_bagger - Spilling /tmp/tmp.oJvWETFBhO/download/blumenbach_anatomie_1805.ocrd.zip to /tmp/tmp.oJvWETFBhO/blumenbach_anatomie_1805
Workspace[remote=False, directory=/tmp/tmp.oJvWETFBhO/blumenbach_anatomie_1805, baseurl=None, file_groups=['OCR-D-GT-SEG-PAGE', 'OCR-D-GT-SEG-BLOCK', 'OCR-D-GT-SEG-LINE', 'OCR-D-IMG'], files=['<OcrdFile fileGrp=OCR-D-GT-SEG-PAGE ID=OCR-D-GT-SEG-PAGE_0001, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-PAGE/OCR-D-GT-SEG-PAGE_0001.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-PAGE ID=OCR-D-GT-SEG-PAGE_0002, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-PAGE/OCR-D-GT-SEG-PAGE_0002.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-PAGE ID=OCR-D-GT-SEG-PAGE_0003, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-PAGE/OCR-D-GT-SEG-PAGE_0003.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-BLOCK ID=OCR-D-GT-SEG-BLOCK_0001, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-BLOCK/OCR-D-GT-SEG-BLOCK_0001.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-BLOCK ID=OCR-D-GT-SEG-BLOCK_0002, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-BLOCK/OCR-D-GT-SEG-BLOCK_0002.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-BLOCK ID=OCR-D-GT-SEG-BLOCK_0003, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-BLOCK/OCR-D-GT-SEG-BLOCK_0003.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-LINE ID=OCR-D-GT-SEG-LINE_0001, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-LINE/OCR-D-GT-SEG-LINE_0001.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-LINE ID=OCR-D-GT-SEG-LINE_0002, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-LINE/OCR-D-GT-SEG-LINE_0002.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-LINE ID=OCR-D-GT-SEG-LINE_0003, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-LINE/OCR-D-GT-SEG-LINE_0003.xml]/> ', '<OcrdFile fileGrp=OCR-D-IMG ID=OCR-D-IMG_0001, mimetype=image/tiff, url=---, local_filename=OCR-D-IMG/OCR-D-IMG_0001.tif]/> ', '<OcrdFile fileGrp=OCR-D-IMG ID=OCR-D-IMG_0002, mimetype=image/tiff, url=---, local_filename=OCR-D-IMG/OCR-D-IMG_0002.tif]/> ', '<OcrdFile fileGrp=OCR-D-IMG ID=OCR-D-IMG_0003, mimetype=image/tiff, url=---, local_filename=OCR-D-IMG/OCR-D-IMG_0003.tif]/> ']]
/tmp/tmp.oJvWETFBhO/blumenbach_anatomie_1805 ~/project
~/project
/tmp/tmp.oJvWETFBhO/blumenbach_anatomie_1805 ~/project
~/project
bash tests/run_alignment_test.bash 
--2024-02-13 14:12:46--  http://hdl.handle.net/21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780
Resolving hdl.handle.net (hdl.handle.net)... 134.76.30.195, 54.163.52.124, 52.19.202.222, ...
Connecting to hdl.handle.net (hdl.handle.net)|134.76.30.195|:80... connected.
HTTP request sent, awaiting response... 302 
Location: http://ola-hd.ocr-d.de/export?id=21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780&internalId=false [following]
--2024-02-13 14:12:46--  http://ola-hd.ocr-d.de/export?id=21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780&internalId=false
Resolving ola-hd.ocr-d.de (ola-hd.ocr-d.de)... 141.5.99.45
Connecting to ola-hd.ocr-d.de (ola-hd.ocr-d.de)|141.5.99.45|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ola-hd.ocr-d.de/export?id=21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780&internalId=false [following]
--2024-02-13 14:12:47--  https://ola-hd.ocr-d.de/export?id=21.11156/6B119B3C-A24A-424C-AC3C-27E64B051780&internalId=false
Connecting to ola-hd.ocr-d.de (ola-hd.ocr-d.de)|141.5.99.45|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘/tmp/tmp.p6yMwcYAvW/download/blumenbach_anatomie_1805.ocrd.zip’

/tmp/tmp.p6yMwcYAvW     [                <=> ]  41.18M  12.3MB/s    in 3.3s    

2024-02-13 14:12:51 (12.3 MB/s) - ‘/tmp/tmp.p6yMwcYAvW/download/blumenbach_anatomie_1805.ocrd.zip’ saved [43176289]

14:12:52.056 INFO ocrd.workspace_bagger - Spilling /tmp/tmp.p6yMwcYAvW/download/blumenbach_anatomie_1805.ocrd.zip to /tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805
Workspace[remote=False, directory=/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805, baseurl=None, file_groups=['OCR-D-GT-SEG-PAGE', 'OCR-D-GT-SEG-BLOCK', 'OCR-D-GT-SEG-LINE', 'OCR-D-IMG'], files=['<OcrdFile fileGrp=OCR-D-GT-SEG-PAGE ID=OCR-D-GT-SEG-PAGE_0001, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-PAGE/OCR-D-GT-SEG-PAGE_0001.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-PAGE ID=OCR-D-GT-SEG-PAGE_0002, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-PAGE/OCR-D-GT-SEG-PAGE_0002.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-PAGE ID=OCR-D-GT-SEG-PAGE_0003, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-PAGE/OCR-D-GT-SEG-PAGE_0003.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-BLOCK ID=OCR-D-GT-SEG-BLOCK_0001, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-BLOCK/OCR-D-GT-SEG-BLOCK_0001.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-BLOCK ID=OCR-D-GT-SEG-BLOCK_0002, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-BLOCK/OCR-D-GT-SEG-BLOCK_0002.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-BLOCK ID=OCR-D-GT-SEG-BLOCK_0003, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-BLOCK/OCR-D-GT-SEG-BLOCK_0003.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-LINE ID=OCR-D-GT-SEG-LINE_0001, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-LINE/OCR-D-GT-SEG-LINE_0001.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-LINE ID=OCR-D-GT-SEG-LINE_0002, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-LINE/OCR-D-GT-SEG-LINE_0002.xml]/> ', '<OcrdFile fileGrp=OCR-D-GT-SEG-LINE ID=OCR-D-GT-SEG-LINE_0003, mimetype=application/vnd.prima.page+xml, url=---, local_filename=OCR-D-GT-SEG-LINE/OCR-D-GT-SEG-LINE_0003.xml]/> ', '<OcrdFile fileGrp=OCR-D-IMG ID=OCR-D-IMG_0001, mimetype=image/tiff, url=---, local_filename=OCR-D-IMG/OCR-D-IMG_0001.tif]/> ', '<OcrdFile fileGrp=OCR-D-IMG ID=OCR-D-IMG_0002, mimetype=image/tiff, url=---, local_filename=OCR-D-IMG/OCR-D-IMG_0002.tif]/> ', '<OcrdFile fileGrp=OCR-D-IMG ID=OCR-D-IMG_0003, mimetype=image/tiff, url=---, local_filename=OCR-D-IMG/OCR-D-IMG_0003.tif]/> ']]
/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805 ~/project
~/project
14:12:56.485 INFO ocrd.cli.resmgr - Downloading registered resource 'fraktur.pyrnn.gz' (https://github.com/zuphilip/ocropy-models/raw/master/fraktur.pyrnn.gz)
  [------------------------------------]    0%14:12:58.261 INFO ocrd.resource_manager._download_impl - Downloading https://github.com/zuphilip/ocropy-models/raw/master/fraktur.pyrnn.gz to /root/.local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur.pyrnn.gz
  [####################################]  100%
14:12:59.051 INFO ocrd.cli.resmgr - Installed resource https://github.com/zuphilip/ocropy-models/raw/master/fraktur.pyrnn.gz under /root/.local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur.pyrnn.gz
14:12:59.051 INFO ocrd.cli.resmgr - Use in parameters as 'fraktur.pyrnn.gz'
14:13:02.195 INFO ocrd.cli.resmgr - Downloading registered resource 'fraktur-jze.pyrnn.gz' (https://github.com/jze/ocropus-model_fraktur/raw/master/fraktur.pyrnn.gz)
  [------------------------------------]    0%14:13:04.130 INFO ocrd.resource_manager._download_impl - Downloading https://github.com/jze/ocropus-model_fraktur/raw/master/fraktur.pyrnn.gz to /root/.local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur-jze.pyrnn.gz
  [####################################]  100%
14:13:04.621 INFO ocrd.cli.resmgr - Installed resource https://github.com/jze/ocropus-model_fraktur/raw/master/fraktur.pyrnn.gz under /root/.local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur-jze.pyrnn.gz
14:13:04.621 INFO ocrd.cli.resmgr - Use in parameters as 'fraktur-jze.pyrnn.gz'
14:13:06.495 DEBUG ocrd.resolver.workspace_from_url - workspace_from_url
mets_basename='mets.xml'
mets_url='/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805/mets.xml'
src_baseurl='/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805'
dst_dir='/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805'
14:13:06.495 DEBUG ocrd.resolver.download_to_directory - directory=|/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805| url=|/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805/mets.xml| basename=|mets.xml| if_exists=|skip| subdir=|None|
14:13:06.495 DEBUG ocrd.resolver.download_to_directory - Stop early, src_path and dst_path are the same: '/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805/mets.xml' (url: '/tmp/tmp.p6yMwcYAvW/blumenbach_anatomie_1805/mets.xml')
14:13:06.496 DEBUG ocrd.workspace_validator - input_file_grp=['OCR-D-GT-SEG-LINE'] output_file_grp=['OCR-D-CIS-OCR-1']
14:13:06.496 DEBUG ocrd.processor.helpers.run_processor - Running processor <class 'ocrd_cis.ocropy.recognize.OcropyRecognize'>
14:13:06.497 DEBUG ocrd.ocrd-cis-ocropy-recognize.resolve_resource - Resolved fraktur.pyrnn.gz to absolute path /root/.local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur.pyrnn.gz
Traceback (most recent call last):
  File "/usr/local/bin/ocrd-cis-ocropy-recognize", line 8, in <module>
    sys.exit(ocrd_cis_ocropy_recognize())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/ocrd_cis/ocropy/cli.py", line 48, in ocrd_cis_ocropy_recognize
    return ocrd_cli_wrap_processor(OcropyRecognize, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/ocrd/decorators/__init__.py", line 133, in ocrd_cli_wrap_processor
    run_processor(processorClass, mets_url=mets, workspace=workspace, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/ocrd/processor/helpers.py", line 87, in run_processor
    processor = get_processor(
  File "/usr/local/lib/python3.8/site-packages/ocrd/processor/helpers.py", line 401, in get_processor
    return processor_class(
  File "/usr/local/lib/python3.8/site-packages/ocrd_cis/ocropy/recognize.py", line 93, in __init__
    self.setup()
  File "/usr/local/lib/python3.8/site-packages/ocrd_cis/ocropy/recognize.py", line 98, in setup
    self.network = load_object(self.get_model(), verbose=1)
  File "/usr/local/lib/python3.8/site-packages/ocrd_cis/ocropy/ocrolib/common.py", line 453, in load_object
    return pickle.load(stream, encoding='latin1')
ModuleNotFoundError: No module named 'ocrolib'
make: *** [Makefile:24: tests/run_alignment_test.bash] Error 1

Exited with code exit status 2
```

Reply:
Thanks!

Looks very strange – never seen that one, and there seems to be an inconsistency between code and actual error.

I'll try to reproduce locally.

Reply:
I now get an HTTP 500 for the hdl.handle.net URL...

Reply:
I managed to convince @tboenig to publish the OCR-D GT under the new Github regime on https://github.com/OCR-D/gt_structure_text/releases. This allows specifying correct dataset URLs.

I also found out what was wrong with the Ocropy recognition models: the pickler did not resolve `ocrolib` in sys.path, because ocrd_cis wraps it under `ocrd_cis.ocropy`.

Alas, the GT bags are [defunct](https://github.com/OCR-D/core/issues/1189) – this must be solved first.

Reply:
It's been a deep dive – lots of progress, but I now have problems with the postcorrection itself: https://github.com/cisocrgroup/ocrd-postcorrection/issues/13

Reply:
I could salvage all tests except for the training, which yields:
```
23:56:39.488 DEBUG cis.CommandLineArguments - current log level: DEBUG
23:56:39.491 DEBUG cis.CommandLineArguments - reading parameters from file /dev/fd/63
23:56:40.66 DEBUG cis.CommandLineArguments - created dir /tmp/tmp.sAyMidd1HJ/train
23:56:40.67 DEBUG cis.CommandLineArguments - created dir /tmp/tmp.sAyMidd1HJ/train/cache
23:56:40.67 DEBUG cis.TrainCommand - command: cis.TrainCommand
23:56:41.199 INFO cis.TrainCommand - le/rr training for OCR-D-CIS-ALIGN(1)
23:56:41.200 DEBUG cis.TrainCommand - calculating profile for ifg=OCR-D-CIS-ALIGN
23:56:41.201 DEBUG cis.TrainCommand - reading base ocr tokens for input file group OCR-D-CIS-ALIGN
23:56:41.293 DEBUG cis.TrainCommand - loading page
23:56:41.295 DEBUG cis.TrainCommand - opening relative file path /tmp/tmp.sAyMidd1HJ/blumenbach_anatomie_1805/OCR-D-CIS-ALIGN/OCR-D-CIS-ALIGN_0001.xml
23:56:41.337 DEBUG cis.TrainCommand - loading page
23:56:41.338 DEBUG cis.TrainCommand - opening relative file path /tmp/tmp.sAyMidd1HJ/blumenbach_anatomie_1805/OCR-D-CIS-ALIGN/OCR-D-CIS-ALIGN_0002.xml
23:56:41.356 DEBUG cis.TrainCommand - loading page
23:56:41.358 DEBUG cis.TrainCommand - opening relative file path /tmp/tmp.sAyMidd1HJ/blumenbach_anatomie_1805/OCR-D-CIS-ALIGN/OCR-D-CIS-ALIGN_0003.xml
23:56:43.144 INFO cis.TrainCommand - read 431 base ocr tokens for input file group OCR-D-CIS-ALIGN
23:56:43.146 DEBUG cis.TrainCommand - profiler command: /tmp/tmp.sAyMidd1HJ/bin/profiler.bash --sourceFormat EXT --config /path/to/language.ini --sourceFile /dev/stdin --jsonOutput /dev/stdout --types
23:56:43.169 DEBUG cis.TrainCommand - caching profile to /tmp/tmp.sAyMidd1HJ/train/cache/OCR-D-CIS-ALIGN.json.gz
23:56:43.179 DEBUG cis.TrainCommand - preparing le-1
23:56:43.191 DEBUG cis.TrainCommand - preparing rr-1
23:56:43.191 INFO cis.TrainCommand - train(/tmp/tmp.sAyMidd1HJ/train/le_training_1.arff, /tmp/tmp.sAyMidd1HJ/train/le_model_1.bin)
23:56:43.196 DEBUG cis.TrainCommand - training logistic classifier from file /tmp/tmp.sAyMidd1HJ/train/le_training_1.arff
weka.core.WekaException: weka.classifiers.functions.Logistic: Not enough training instances with class labels (required: 1, provided: 0)!
	at weka.core.Capabilities.test(Capabilities.java:1332)
	at weka.core.Capabilities.test(Capabilities.java:1208)
	at weka.core.Capabilities.testWithFail(Capabilities.java:1506)
	at weka.classifiers.functions.Logistic.buildClassifier(Logistic.java:678)
	at de.lmu.cis.ocrd.ml.LogisticClassifier.train(LogisticClassifier.java:43)
	at de.lmu.cis.ocrd.ml.Trainer.train(Trainer.java:79)
	at de.lmu.cis.ocrd.cli.TrainCommand.execute(TrainCommand.java:40)
	at de.lmu.cis.ocrd.cli.Main.run(Main.java:33)
	at de.lmu.cis.ocrd.cli.Main.main(Main.java:9)
```

I have no idea what to do about this, so I simply deactivated that test.